sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -324,6 +325,124 @@ object LogicalPlanIntegrity {
       LogicalPlanIntegrity.hasUniqueExprIdsForOutput(plan))
   }
 
+  /**
+   * This method validates there are no dangling attribute references.
+   * Returns an error message if the check does not pass, or None if it does pass.
+   */
+  def validateNoDanglingReferences(plan: LogicalPlan): Option[String] = {
+    plan.collectFirst {
+      // DML commands and multi instance relations (like InMemoryRelation caches)
+      // have different output semantics than typical queries.
+      case _: Command => None
+      case _: MultiInstanceRelation => None
+      case n if canGetOutputAttrs(n) =>
+        val inputExprIds = (n.children.flatMap(c => c.output ++ c.metadataOutput) ++ n.output)
+          .map(_.exprId).toSet
+        val danglingReferences = n.references.filter {
+          a => a.resolved && !inputExprIds.contains(a.exprId)
+        }.map(_.qualifiedName)
+        if (danglingReferences.nonEmpty) {
+          Some(s"Aliases ${danglingReferences.mkString(", ")} are dangling " +
+            s"in the references for plan:\n ${n.treeString}")
+        } else {
+          None
+        }
+    }.flatten
+  }
+
+  /**
+   * Validate that the grouping key types in Aggregate plans are valid.
+   * Returns an error message if the check fails, or None if it succeeds.
+   */
+  def validateGroupByTypes(plan: LogicalPlan): Option[String] = {
+    plan.collectFirst {
+      case a @ Aggregate(groupingExprs, _, _) =>
+        val badExprs = groupingExprs.filter(_.dataType.isInstanceOf[MapType]).map(_.toString)
+        if (badExprs.nonEmpty) {
+          Some(s"Grouping expressions ${badExprs.mkString(", ")} cannot be of type Map " +
+            s"for plan:\n ${a.treeString}")
+        } else {
+          None
+        }
+    }.flatten
+  }
+
+  /**
+   * Validate that the aggregation expressions in Aggregate plans are valid.
+   * Returns an error message if the check fails, or None if it succeeds.
+   */
+  def validateAggregateExpressions(plan: LogicalPlan): Option[String] = {
+    /**
+    * Returns true if all the exprIds, referenced in an Expression <subExpr>
+    * (used in the aggregate expressions of an Aggregate),
+    * are either used as measures (under an AggregateFunction) or are explicitly grouped,
+    * given the <groupingExprIds> of all grouping expressions
+    * (used in the grouping expressions of the same Aggregate)
+
+    * I.e:
+    * <groupingExprIds> can be used "freely" in an AggregateExpressions, all other exprIds can only
+    * be used under an AggregateFunction (i.e. aggregated).
+     */
+    def isAggregateSubExpressionValid(
```

Comment:
is this trying to match https://github.com/apache/spark/blob/branch-3.4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L405-L476 ?

Reply:
in a sense they are similar! Eventually we can explore how to merge analysis validation with per-rule validation. BTW this validation deals with issues we've seen during rewrites specifically, where the exprids before and after the rule are messed up.